### PR TITLE
Omar

### PR DESCRIPTION
**Pull request by ob2301**
Implementing FordA dataset (binary classification) with MOMENT backbone.

### CODE REVIEW

File: src/fmtk/datasets/forda.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import os
+
+class FordADataset(TimeSeriesDataset):
+    def __init__(
+        self,
+        dataset_cfg,
+        task_cfg,
+        split,
+        data_stride_len: int = 1,
+        random_seed: int = 13,
```
Comment (krishna-praneet) [inline src/fmtk/datasets/forda.py]: Can we add `data_stride_len` and `random_seed` arguments to the `dataset_cfg` itself? You could then use them like `self.data_stride_len = dataset_cfg["data_stride_len"]`. This would help maintain the convention of dataset constructor signatures, which take only three arguments: dataset_cfg, task_cfg, and split.

File: src/fmtk/datasets/forda.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        #fit scaler only on training data
+        if self.split == "train":
+            self.scaler = StandardScaler()
+            self.scaler.fit(X)
+            np.save(mean_path, self.scaler.mean_)
+            np.save(scale_path, self.scaler.scale_)
+        else:
+            mean = np.load(mean_path)
+            scale = np.load(scale_path)
```
Comment (krishna-praneet) [inline src/fmtk/datasets/forda.py:56]: If someone called only the `test` split without ever calling train first, this would crash. Could we add some checks to see if the files exist and then compute them if they don't ?

**Review by krishna-praneet — CHANGES_REQUESTED** [review]

File: src/fmtk/components/backbones/mantis.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+from fmtk.components.base import BaseModel
+from mantis.architecture import Mantis8M
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py:2]: Are you using `mantis-tsfm` library, could you please add that to the dependencies as well in `pyproject.toml` file.

File: src/fmtk/components/backbones/mantis.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        # changed this to match Chronos args (just pass batch)
+        batch = batch_x
+        
+        x, mask, y = self.preprocess(batch)
+
+        # x_np = x.cpu().numpy()
+        # emb_np = self.model.transform(x_np)
+        # embedding = torch.tensor(emb_np, dtype=torch.float32, device=self.device)
+
+        self.network.to(self.device)
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: Let's remove this, we are handling the device in the constructor function.

File: src/fmtk/components/backbones/mantis.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class MantisModel(BaseModel):
+    def __init__(self, device, model_name="8M"):
+        super().__init__()
+        self.device = device
+        self.peft_enable = False
+
+        # initialization step
+        self.network = Mantis8M(device="cpu")
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: Let's directly initialize on the device requested instead of always on cpu first. 
`Mantis8M(device = self.device)`

File: src/fmtk/components/backbones/mantis.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        self.network = Mantis8M.from_pretrained(f"paris-noah/Mantis-{model_name}")
+
+        # next step is wrapping it with trainer
+        self.model = MantisTrainer(device="cpu", network=self.network)
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: `device = self.device`

File: src/fmtk/components/backbones/mantis.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+from mantis.trainer import MantisTrainer
+import torch
+import numpy as np
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from peft import LoraConfig, get_peft_model
+
+
+class MantisModel(BaseModel):
+    def __init__(self, device, model_name="8M"):
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: Let's keep the class constructor signature the same as others `(self, device, model_name="8M", model_config=None)`. This way, if someone sends `model_config` inappropriately, it will handle it gracefully.

File: src/fmtk/components/backbones/mantis.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        self.network = Mantis8M(device="cpu")
+        self.network = Mantis8M.from_pretrained(f"paris-noah/Mantis-{model_name}")
+
+        # next step is wrapping it with trainer
+        self.model = MantisTrainer(device="cpu", network=self.network)
+        self.peft_enable = False
+
+    # preprocess step almost the same as MOMENTs since both expect (batch, channels, length),
+    # and they both need float inputs
+    def preprocess(self, batch):
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: There has been an update to `preprocess`. Please refer to the moment or the chronos file. The `batch_x` sent to `preprocess` will not contain `y`. We realized that it was unnecessary for us to send to `y` through `forward` and `preprocess`.  So, to maintain consistency, 
- Could you please add the function signature of `preprocess(self, batch_x, mask=None)`
- Remove the usage of `y` in preprocess.

File: src/fmtk/components/backbones/mantis.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+        # Mantis expects 1 input channel average multiple channels if present
+        if x.shape[1] != 1:
+            x = x.mean(dim=1, keepdim=True)
+
+        return x, mask, y
+
+    def forward(self, batch_x, mask=None):
+        # changed this to match Chronos args (just pass batch)
+        batch = batch_x
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: Same here. We are removing `y` from `forward` function.

File: src/fmtk/components/backbones/mantis.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        
+        x, mask, y = self.preprocess(batch)
+
+        # x_np = x.cpu().numpy()
+        # emb_np = self.model.transform(x_np)
+        # embedding = torch.tensor(emb_np, dtype=torch.float32, device=self.device)
+
+        self.network.to(self.device)
+        self.network.train()  # enable training mode if needed
```
Comment (krishna-praneet) [inline src/fmtk/components/backbones/mantis.py]: Let us move line 72 inside the `enable_peft()` method. We are not allowing full finetuning of the backbone yet, even then we would like to keep this away from the `forward` function, because we will use the forward for both training and evaluation. So if we do just an evaluation, we would not want the overhead from instantiating the optimizer states and gradients (which are only required during training).

**Review by krishna-praneet — CHANGES_REQUESTED** [review]
Omar, great work. Some crucial changes, then we can make this available on the FMTK.

**Comment by krishna-praneet** [comment]
Setting this as duplicate as main branch has been updated. Changes have been rebased and  merged through PR #7.
